File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    int getMinJump(int i, vector<int> &nums, vector<int> &dp){
        if(i>=nums.size()-1) return 0;
        if(dp[i]!=-1) return dp[i];
        
        int minJump=nums.size();
        
        for(int index=i+1; index<=i+nums[i]; index++){
            minJump=min(minJump, 1+getMinJump(index, nums, dp));
        }
        return dp[i]=minJump;
    }
    
    int jump(vector<int>& nums) {
        vector<int> dp(nums.size(), -1);
        return getMinJump(0, nums, dp);
    }
};

File: 0045-jump-game-ii/0045-jump-game-ii.cs
public class Solution {
    
    int solve(int ind, int[] nums, int[] dp, int n)
    {
        if(ind>=n-1)return 0;
        if(dp[ind]!=-1)return dp[ind];
        
        int mini = n;  //int.MaxValue dile jame ace overflow asbe
        
        for(int i=ind+1; i<=ind+nums[ind]; i++)
        {
            mini = Math.Min(mini, 1+solve(i, nums, dp, n));
        }
        
        return dp[ind]=mini;
    }
    
    public int Jump(int[] nums) {
        
        int n = nums.Length;
        int[] dp = new int[n];
        
        for(int i=0; i<n; i++)dp[i]=-1;
        
        return solve(0, nums, dp, n);
    }
}